projects/clr/hipamd/src/hip_comgr_helper.cpp

```diff
@@ -528,39 +532,57 @@ bool linkLLVMBitcode(const amd_comgr_data_set_t linkInputs, const std::string& i
                      std::vector<std::string>& linkOptions, std::string& buildLog,
                      std::vector<char>& LinkedLLVMBitcode) {
   const amd_comgr_language_t lang = AMD_COMGR_LANGUAGE_HIP;
-  amd_comgr_action_info_t action;
+  ComgrActionInfoUniqueHandle LinkAction;
 
-  if (!createAction(action, linkOptions, isa, lang)) {
+  if (LinkAction.Create() != AMD_COMGR_STATUS_SUCCESS) {
     return false;
   }
 
-  amd_comgr_data_set_t output;
-  if (auto res = amd::Comgr::create_data_set(&output); res != AMD_COMGR_STATUS_SUCCESS) {
-    amd::Comgr::destroy_action_info(action);
+  ComgrDataSetUniqueHandle output;
+  if (output.Create() != AMD_COMGR_STATUS_SUCCESS) {
     return false;
   }
 
-  if (auto res = amd::Comgr::do_action(AMD_COMGR_ACTION_LINK_BC_TO_BC, action, linkInputs, output);
-      res != AMD_COMGR_STATUS_SUCCESS) {
-    amd::Comgr::destroy_action_info(action);
-    amd::Comgr::destroy_data_set(output);
+  // If inputs contain BC_BUNDLE, unbundle first to BC; else link inputs as-is.
+  size_t bundleCount = 0;
+  auto res = amd::Comgr::action_data_count(linkInputs, AMD_COMGR_DATA_KIND_BC_BUNDLE, &bundleCount);
+  if (res != AMD_COMGR_STATUS_SUCCESS) {
     return false;
   }
 
-  if (!extractBuildLog(output, buildLog)) {
-    amd::Comgr::destroy_action_info(action);
-    amd::Comgr::destroy_data_set(output);
+  if (bundleCount > 0) {
+    ComgrDataSetUniqueHandle unbundleInputs;
+    if (unbundleInputs.Create() != AMD_COMGR_STATUS_SUCCESS) {
+      return false;
+    }
+    res = amd::Comgr::do_action(AMD_COMGR_ACTION_UNBUNDLE, LinkAction.get(), linkInputs, unbundleInputs.get());
+    if (res != AMD_COMGR_STATUS_SUCCESS) {
```

Comment:
Should we clean up all resources before returning? Probably we should have a single point of return and before returning we should check all the open COMGR resources and close them?